core/src/config/workflow.ts

```diff
@@ -255,21 +256,27 @@ export const workflowStepSchema = createSchema({
 
       \`onSuccess\` (default): This step will be run if all preceding steps succeeded or were skipped.
 
-      \`onError\`: This step will be run if a preceding step failed, or if its preceding step has \`when: onError\`.
+      \`onError\`: This step will be run if a preceding step throws any error, or if its preceding step has \`when: onError\`.
       If the next step has \`when: onError\`, it will also be run. Otherwise, all subsequent steps are ignored.
 
+      \`onFailure\`: This step will be run if a preceding step throws an error with \`allowFailure: false\`, or if its preceding step has \`when: onFailure\`.
+      If the next step has \`when: onFailure\`, it will also be run. Otherwise, all subsequent steps are ignored.
+
       \`always\`: This step will always be run, regardless of whether any preceding steps have failed.
 
       \`never\`: This step will always be ignored.
 
       See the [workflows guide](${DOCS_BASE_URL}/using-garden/workflows#the-skip-and-when-options) for details
       and examples.
       `),
+    allowFailure: joi
+      .boolean()
```

Comment:
We should add a default: 
```suggestion
      .boolean()
      .default(false)
```